rfcs/0042-bitswap.md

```diff
@@ -0,0 +1,268 @@
+
+  module Sync_state = struct
+    (* global sync state *)
+    let sync_table = String.Table.create ()
+
+    (* runs a continuous background task to synchronize resources tracked in the daemon with the Bitswap cache *)
+    let run_sync_task () =
+
+    (* track a new resource in global state *)
+    let track ((module Resource) : 'a advertisable) (resource : 'a) =
+      ...
+
+    (* mark a resource currently tracked in global state to be processed for collection in a future async cycle *)
+    let mark_for_collection ((module Resource) : 'a advertisable) (resource : 'a) =
+      ...
+  end
+
+  module Bitswap_advertised (Resource : Bitswap_advertisable_intf) = struct
+    type t = Resource.t
+
+    let create init_args =
+      let resource = Resource.create init_args in
+      Sync_state.add (module Resource) x ;
+      Gc.Expert.attach_finalizer x (fun _ -> Sync_state.mark_for_removal (module Resource) x) ;
+      x
+  end
+end
+```
+
+### Libp2p Helper IPC Modifications
+
+New IPC messages are required in order to support the daemon requesting resources stored in Bitswap blocks from the helper process. When the daemon needs to acquire a resource that is stored in Bitswap, it will send a message to the libp2p helper process asking it to download all the blocks associated with that resource (identified by the root Bitswap block CID). The helper process will begin the process of downloading all Bitswap blocks in the tree pointed to by the root Bitswap block, and will inform the helper upon success or failure of that operation.
+
+Whenever new Bitswap blocks are available to the node, the libp2p helper needs to mark an entry in the DHT saying that our node can provide this block of data. When the blocks in question are downloaded from another peer in the network, this is done [automatically by the libp2p Bitswap implementation](https://github.com/ipfs/go-bitswap/blob/0fa397581ca6197c6c4ca17842719370f6596e95/bitswap.go#L254). However, we still need to have custom logic for adding new DHT provision keys for data we broadcast from out node locally, and when a node restarts from with a pre-existing Bitswap block cache.
+
+In order to support this, whenever we submit [new pub/sub broadcast messages from the daemon](https://github.com/MinaProtocol/mina/blob/develop/src/app/libp2p_helper/src/libp2p_helper/main.go#L452), the daemon will include a list of Bitswap block CIDs associated with any extraneous data related to the broadcast message. The libp2p helper process will mark these in the DHT.
+
+We do not need to worry about advertising persisted block CIDs on restarts so long as the node is restarted with the same libp2p keypair, as the DHT records informing what block CIDs can be served by this node will persist on the network after the node goes offline. We also don't need to concern ourselves with pruning garbage collected block CIDs from the DHT since all provide keys we store there only persist for 24 hours.
```

Comment:
> We should adjust the daemon behavior to reuse libp2p keypairs if we haven't already (have we?). I think we can just store the generated keypair in the mina_config directory and look there on startup if nothing is passed via CLI

From what I can tell in the code, we do already have this behavior in our daemon. Though, I think our seeds don't have persistence for the `.mina-config` directory, so we would want to enable that.

Reply:
> Assuming we garbage collect when the blocks leave the transition frontier, doesn't this mean that roughly a third of the DHT records will point to missing blocks? (290 blocks in the frontier should cycle out every ~1200 minutes). Is this okay?

The DHT implementation doesn't appear to support removing DHT records, so we would need to ask for them to add support for that. This stated, I don't really for-see this being an issue, as we only fallback to the DHT when none of our neighbors have the block in question. Realistically, this will only happen for newly broadcast blocks that are not consistent throughout the network, not for older blocks that nodes are downloading during bootstrap or catchup.